File: Pepcoding Classes/Foundation/Recursion/l003recursion.java
import java.util.Scanner;
import java.util.ArrayList;

public class l003recursion{
    public static Scanner scn = new Scanner(System.in);

    public static void main(String[] args) throws Exception {
        String str = scn.next();
        
        System.out.println(gss(str,0));
    }



    // IMPORTANT QUESTION   
    // Mobile KeyPad Problem 
    // 0 -> .;
    // 1 -> abc
    // 2 -> def
    // 3 -> ghi
    // 4 -> jkl
    // 5 -> mno
    // 6 -> pqrs
    // 7 -> tu
    // 8 -> vwx
    // 9 -> yz
    
    static String[] codes = {".;","abc","def","ghi","jkl","mno","pqrs","tu","vwx","yz"};

    public static ArrayList < String > getKPC(String str, int idx) {
        if(idx == str.length()){
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }
        
        ArrayList<String> recAns = getKPC(str,idx + 1);
        ArrayList<String> myAns = new ArrayList<>();
        
        char ch = str.charAt(idx);
        ArrayList<String> ans = new ArrayList<>();
        
        String code = codes[ch - '0'];
        for(int i = 0; i < code.length() ; i++){
            char c = code.charAt(i);
            
            for(String s : recAns){
                myAns.add(c + s);
            }
        }

        return myAns;
    }

// ===================================================================================

    // Get Subsequence of a String 
    // Example :- 
    // Input : "abc"
    // Output : [, c, b, bc, a, ac, ab, abc]

    public static ArrayList < String > gss(String str, int idx) {  
        if(idx == str.length()){
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }
        
        char ch = str.charAt(idx);
        
        ArrayList<String> recAns = gss(str, idx +1);
        ArrayList<String> myAns = new ArrayList<>(recAns);
        
        for (String s : recAns){
            myAns.add(ch + s);
        }
        
        
        return myAns;
    }

    // ALL TYPES OF SUBSEQUENCE USING RETURN TYPE & WAY UP 
    public static void printSS(String str, int idx, String ans) {
        if (idx == str.length()) {
            System.out.println(ans);
            return;
        }

        printSS(str, idx + 1, ans + str.charAt(idx));
        printSS(str, idx + 1, ans);
    }

    public static void printSS_02(String str, int idx, StringBuilder ans) {
        if (idx == str.length()) {
            System.out.println(ans);
            return;
        }

        ans.append(str.charAt(idx));
        printSS_02(str, idx + 1, ans);
        ans.deleteCharAt(ans.length() - 1);

        printSS_02(str, idx + 1, ans);
    }

    // return Type
    public static ArrayList<String> subseq_03(String str, int idx) {
        if (idx == str.length()) {
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }

        ArrayList<String> recAns = subseq_03(str, idx + 1);

        ArrayList<String> myAns = new ArrayList<>(recAns); // nahi anne ki choise dekhli.

        char ch = str.charAt(idx);
        for (String s : recAns) {
            myAns.add(ch + s);
        }

        return myAns;
    }

    public static void printASCIISS(String str, int idx, String ans) {
        if (idx == str.length()) {
            System.out.println(ans);
            return;
        }

        printASCIISS(str, idx + 1, ans + str.charAt(idx));
        printASCIISS(str, idx + 1, ans + (int) str.charAt(idx));
        printASCIISS(str, idx + 1, ans);
    }

    public static ArrayList<String> printASCIISS_Ret(String str, int idx) {
        if (idx == str.length()) {
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }

        ArrayList<String> recAns = printASCIISS_Ret(str, idx + 1);

        ArrayList<String> myAns = new ArrayList<>(); // nahi anne ki choise dekhli.

        char ch = str.charAt(idx);
        for (String s : recAns) {
            myAns.add(ch + s);
            myAns.add((int) ch + s);
            myAns.add(s);
        }

        return myAns;
    }

}

File: Pepcoding Classes/Foundation/Recursion/l003recursion.cpp
#include<iostream>
#include<vector>

using namespace std;

    // IMPORTANT QUESTION   
    // Mobile KeyPad Problem 
    // 0 -> .;
    // 1 -> abc
    // 2 -> def
    // 3 -> ghi
    // 4 -> jkl
    // 5 -> mno
    // 6 -> pqrs
    // 7 -> tu
    // 8 -> vwx
    // 9 -> yz
    
static vector<string> codes = {".;","abc","def","ghi","jkl","mno","pqrs","tu","vwx","yz"};

vector<string> getKPC(string str, int idx) {
    if(idx == str.length()){
        vector<string> base;
        base.push_back("");
        return base;
    }
    
    vector<string> recAns = getKPC(str,idx + 1);
    vector<string> myAns;
    
    char ch = str[idx];
    
    string code = codes[ch - '0'];
    for(int i = 0; i < code.length() ; i++){
        char c = code[i];
        
        for(string s : recAns){
            myAns.push_back(c + s);
        }
    }
    

    return myAns;
}

// -------------------------------------------------------------


// Get the Paths of the Maze (0,0) -> (n,m)
vector<string> getMazePaths(int sr, int sc, int dr, int dc){
    if(sr == dr && sc == dc){
        vector<string> base; 
        base.push_back("");
        return base;
    }
    
    vector<string> myAns ;
    if(sc + 1 <= dc){
        vector<string> Horizontal = getMazePaths(sr,sc + 1 , dr, dc);
        for(string s : Horizontal){
            myAns.push_back("h" + s);
        }
    }
    
    if(sr + 1 <= dr){
        vector<string> Vertical = getMazePaths(sr + 1,sc, dr, dc);
        for(string s : Vertical){
            myAns.push_back("v" + s);
        }
    }
    
    return myAns;
}

// ----------------------------------------

// Maze Paths With Jumps 

vector<string> getMazePathsWithJumps(int sr, int sc, int dr, int dc){
    if(sr == dr && sc == dc){
        vector<string> base; 
        base.push_back("");
        return base;
    }
    
    vector<string> myAns ;
    
    for (int i = 1; i <= dc - sc; i++){
        
        vector<string> Horizontal = getMazePathsWithJumps(sr,sc + i , dr, dc);
        for(string s : Horizontal){
            myAns.push_back("h"+ to_string(i) + s);
        }
    }
    
    for (int i = 1; i <= dr - sr; i++){
        
        vector<string> Vertical = getMazePathsWithJumps(sr + i,sc, dr, dc);
        for(string s : Vertical){
            myAns.push_back("v"+ to_string(i) + s);
        }
    }
    
    
    for (int i = 1; i <= ((dr - sr) && i <= (dc - sc)); i++){
        
        vector<string> Diagonal = getMazePathsWithJumps(sr + i,sc + i, dr, dc);
        for(string s : Diagonal){
            myAns.push_back("d"+ to_string(i) + s);
        }
    }
    
    return myAns;
}

// -----------------------------------------

// Print the vector 
void display(vector<string> ans){
    cout << "[";
    for(int i = 0 ; i < ans.size(); i++){
        cout << ans[i] ;
        if( i < ans.size() - 1) cout << ", ";
    }
    cout << "]" << endl ;
}

// ------------------------------------------------------------------------------------------

// Stair Path 
// Path of ladder

vector<string> pathOfLadder(int st , int dest , int jump){
    if(st == dest ){
        vector<string> base;
        base.push_back("");
        return base;
    }

    vector<string> ans;
    for(int move = 1 ; move <= dest - st && move <= jump ; move++){
        vector<string>  smallAns = pathOfLadder(st + move , dest , jump);
        for(string s : smallAns){
            ans.push_back(to_string(move) + s);
        }
    }
    return ans ;
} 

// ------------------------------------------------------------------------------------------

int main(){
    int n,m ; cin >> n >> m ;
    display(pathOfLadder(0,n,m));
    
    // display(getKPC("78",0));

    return 0;
}